### PR TITLE
Add DepthMapAttribute and optional dense depth supervision

### PR DESCRIPTION
**Pull request by fwilliams**
## Summary

Adds a new `DepthMapAttribute` (subclass of `PerImageRasterAttribute`) for per-image depth rasters with scale-aware semantics, and wires optional dense depth supervision into `GaussianSplatReconstruction`. Six files changed, +697 / −5.

### `DepthMapAttribute`

- **`DepthScale.METRIC`** depths track the scene's coordinate system. The linear scale of any subsequent spatial transform is folded into `unit_scale` at attribute level rather than rewriting the rasters on disk. Non-similarity or reflective transforms raise rather than silently produce anisotropically deformed depth — this closes the most likely footgun (running `NormalizeScene` after attaching metric depths).
- **`DepthScale.RELATIVE`** depths (e.g. from a monocular relative-depth estimator) are dimensionless and immune to scene transforms.
- **`DepthMissingPolicy.{NAN,ZERO,SENTINEL}`** controls the validity convention so the loader emits a per-pixel validity mask alongside the depth tensor. Default downsample interpolation is `NEAREST` to avoid bilerping across depth discontinuities.

### Reconstructor wiring

- New config fields `dense_depth_attribute` and `dense_depth_reg`. The loss form is keyed off `DepthMapAttribute.scale`:
  - **METRIC** → masked L1 on absolute depths.
  - **RELATIVE** → scale-and-shift invariant L1 (each side normalized by its own median + MAD, per Ranftl et al., MiDaS).
- `SfmDataset` detects `DepthMapAttribute` and emits both `data[name]` and `data[f"{name}_valid"]`.
- `_needs_depth_render(config)` now ORs `sparse_depth_reg` and `dense_depth_reg` so the renderer emits the depth channel for either path — without this, dense-depth-only configs would silently fail with a `None` depth tensor.

## Test plan

- [x] 20 new unit tests in `TestDepthMapAttribute` cover attribute hooks (filter/select/spatial transform), validity policies, loader formats (16-bit PNG, float NPY, sentinel float NPY), state-dict round-trip, similarity-transform enforcement, and end-to-end `apply_transformation_matrix` scale folding. `cd tests && pytest unit/test_scene_attributes.py` passes all 108 tests.
- [ ] End-to-end training-time exercise with a real scene + sensor depth (METRIC) and monocular depth (RELATIVE) — not yet run in CI.

## Notes

- `batch_size > 1` raises `NotImplementedError` for dense depth, symmetric with the existing sparse-depth path.
- Dense depth supervision is fully decoupled from `sparse_depth_reg` / `has_visible_point_indices` — works on any scene with a `DepthMapAttribute`.

### CODE REVIEW

File: fvdb_reality_capture/radiance_fields/gaussian_splat_reconstruction.py [hunk 1]
```diff
@@ -39,6 +39,41 @@
...
 
 
+def _scale_shift_invariant_l1(
+    pred: torch.Tensor, target: torch.Tensor, valid: torch.Tensor, eps: float = 1e-6
+) -> torch.Tensor:
+    """Scale-and-shift invariant L1 loss between ``pred`` and ``target``, per image.
+
+    Each side is independently normalized by its own median and mean-absolute-deviation
+    over valid pixels (Ranftl et al., MiDaS); the L1 of the residual is then averaged
+    over valid pixels. Restricted to batch size 1; the caller enforces this elsewhere.
```
Comment (swahtz) [inline fvdb_reality_capture/radiance_fields/gaussian_splat_reconstruction.py]: **❓ question:** ‏Why is this L1 loss only valid for batch size 1?  Seems like this could be valid for computing across a batch to have a more generic implementation (even if the caller only uses it on training batches of size 1).

File: fvdb_reality_capture/radiance_fields/gaussian_splat_reconstruction.py [hunk 2]
```diff
@@ -157,6 +192,28 @@ class GaussianSplatReconstructionConfig:
...
+    """
+    Name of a :class:`~fvdb_reality_capture.sfm_scene.DepthMapAttribute` registered on the scene to use as a
+    per-pixel ground-truth depth target. If ``None`` and :attr:`dense_depth_reg` is non-zero, a ``ValueError`` is raised.
+
+    Default: ``None`` (no dense depth supervision).
+    """
+
+    dense_depth_reg: float = 0.0
+    """
+    Weight on the dense depth loss term. The dense loss coexists with :attr:`sparse_depth_reg`; both may be enabled
```
Comment (swahtz) [inline fvdb_reality_capture/radiance_fields/gaussian_splat_reconstruction.py:212]: **💡 suggestion:** ‏Is it worth mentioning that `dense_depth_reg` is going to be regularizing an L1 loss on scene units (so its value will need to scale with the scene units for different scenes) whereas `sparse_depth_reg` is on median-normalized L1 (so doesn't need to change between scenes)?

File: fvdb_reality_capture/sfm_scene/depth_map_attribute.py [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+            raise ValueError(
+                f"DepthMapAttribute(scale=METRIC) requires a similarity (uniform-scale) "
+                f"spatial transform; got singular values {svals}. A non-uniform 3x3 would "
+                f"deform depth anisotropically, which cannot be represented by a single "
+                f"scalar multiplier. Convert the attribute to scale=RELATIVE if you "
+                f"intentionally want it to be invariant to such transforms."
+            )
+        if np.linalg.det(linear) < 0:
+            raise ValueError(
+                "DepthMapAttribute(scale=METRIC) does not support reflective (det<0) " "spatial transforms."
```
Comment (swahtz) [inline fvdb_reality_capture/sfm_scene/depth_map_attribute.py]:
```suggestion
                "DepthMapAttribute(scale=METRIC) does not support reflective (det<0) spatial transforms."
```

Weird string formatting

File: fvdb_reality_capture/sfm_scene/depth_map_attribute.py [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+                coordinate system, or :class:`DepthScale.RELATIVE` for scale-invariant
+                depth predictions (e.g. monocular relative depth).
+            missing_policy: Encoding of invalid pixels (see :class:`DepthMissingPolicy`).
+            invalid_value: The sentinel value used when ``missing_policy == SENTINEL``.
+                Must be ``None`` for the other policies.
+            resize_interpolation: Interpolation mode for downsampling. Defaults to
+                :class:`InterpolationMode.NEAREST` to avoid mixing depths across
+                discontinuities.
+        """
+        super().__init__(paths=paths, resize_interpolation=resize_interpolation)
```
Comment (swahtz) [inline fvdb_reality_capture/sfm_scene/depth_map_attribute.py:115]: **💡 suggestion:** ‏I think we should be more opinionated and provide checking for a combination of `resize_interpolation` that isn't `NEAREST` with a `missing_policy` of `ZERO` or `SENTINEL`.  If there's a valid pixel next to a zero/invalid value and the user specifies bilinear/bicubic interp they're going to trip over the trap and get a blurry valid value with the zero/sentinel.  I think we should throw an exception (or at least a strongly worded WARN) for these combos and suggest they use a NaN policy if they want bilinear/bicubic.

File: fvdb_reality_capture/radiance_fields/gaussian_splat_reconstruction.py (hunk 1 above)
Comment (fwilliams) [reply fvdb_reality_capture/radiance_fields/gaussian_splat_reconstruction.py]: Good call. I made `_scale_shift_invariant_l1` batch-generic — it now normalizes each image over its own valid pixels in a short per-element loop (rather than a jagged tensor; batch sizes here are tiny so the loop is simpler and avoids the JT machinery). Empty elements contribute zero, and the `batch_size==1` assert is gone. Added a test asserting the batched result equals the mean of the per-image losses. The training loop still gates dense depth at `batch_size==1` (symmetric with the sparse path), but the helper is no longer the limiting factor. Fixed in ed1d372.

File: fvdb_reality_capture/radiance_fields/gaussian_splat_reconstruction.py (hunk 2 above)
Comment (fwilliams) [reply fvdb_reality_capture/radiance_fields/gaussian_splat_reconstruction.py:212]: Added a note to the `dense_depth_reg` docstring: for METRIC depth the loss is in scene units (so the weight must be retuned per scene), while for RELATIVE depth it's unitless/median-normalized and transfers across scenes — contrasting with `sparse_depth_reg`, which is always median-normalized. ed1d372.

File: fvdb_reality_capture/sfm_scene/depth_map_attribute.py (hunk 3 above)
Comment (fwilliams) [reply fvdb_reality_capture/sfm_scene/depth_map_attribute.py]: Done — merged into a single literal. ed1d372.

File: fvdb_reality_capture/sfm_scene/depth_map_attribute.py (hunk 4 above)
Comment (fwilliams) [reply fvdb_reality_capture/sfm_scene/depth_map_attribute.py:115]: Agreed, made it opinionated: the constructor now raises if `resize_interpolation` is an averaging mode (AREA/BILINEAR/BICUBIC) combined with a ZERO or SENTINEL policy, since downsampling would bleed the invalid fill value into valid pixels. NaN policy is required for averaging modes (NaNs propagate under interpolation and get re-flagged invalid on reload). Added a test. I went with a hard error rather than a warning since it only blocks the genuinely dangerous combo (the canonical NEAREST + ZERO sensor-depth case is unaffected) — happy to downgrade to a WARN if you'd prefer. ed1d372.

**Review by swahtz — APPROVED** [review]
Looks great, thanks for addressing the notes!
